src/lib.rs

```diff
@@ -1,182 +1,53 @@
-                _ => {
-                    *ifa = (**ifa).ifa_next;
-                    continue;
-                }
-            }
-        }
-
-        Ok(interfaces)
-    }
-}
-
-/// Retrieves the local ip address for the current operative system
-pub fn local_ip() -> Result<IpAddr, Error> {
-    let ifas = find_af_inet()?;
-
-    if cfg!(target_os = "macos") {
-        if let Some((_, ipaddr)) = find_ifa(&ifas, "en0") {
-            return Ok(*ipaddr);
-        }
-    }
-
-    if cfg!(target_os = "linux") {
-        if let Some((_, ipaddr)) = find_ifa(&ifas, "wlp2s0") {
-            return Ok(*ipaddr);
-        }
-    }
-
-    Err(Error::PlatformNotSupported(env::consts::OS.to_string()))
-}
-
-/// Finds the network interface with the provided name in the vector of network
-/// interfaces provided
-pub fn find_ifa<'a>(
-    ifas: &'a [(&'a str, IpAddr)],
-    ifa_name: &str,
-) -> Option<&'a (&'a str, IpAddr)> {
-    ifas.iter()
-        .find(|(name, ipaddr)| *name == ifa_name && matches!(ipaddr, IpAddr::V4(_)))
-}
-
-/// Retrieves the name of a interface address
-unsafe fn get_ifa_name<'a>(ifa: *mut *mut ifaddrs) -> Result<&'a str, Error> {
-    CStr::from_ptr((**ifa).ifa_name)
-        .to_str()
-        .map_err(Error::IntAddrNameParseError)
-}
+//! `local-ip-address` is a wrapper on `getifaddrs` which retrieves host's
+//! network interfaces.
+//!
+//! Handy functions are provided such as `local_ip` which retrieve the local IP
+//! address based on the host system
+//!
+//! ```
+//! use std::net::IpAddr;
+//! use local_ip_address::local_ip;
+//!
+//! assert!(matches!(local_ip().unwrap(), IpAddr));
+//! ```
+//!
+//! You are able to iterate over a vector of tuples where the first element of
+//! the tuple is the name of the network interface and the second is the IP
+//! address.
+//!
+//! ```
+//! use std::net::IpAddr;
+//! use local_ip_address::find_af_inet;
+//!
+//! let ifas = find_af_inet().unwrap();
+//!
+//! if let Some((_, ipaddr)) = ifas
+//! .iter()
+//! .find(|(name, ipaddr)| *name == "en0" && matches!(ipaddr, IpAddr::V4(_))) {
+//!     println!("This is your local IP address: {:?}", ipaddr);
+//!     // This is your local IP address: 192.168.1.111
+//!     assert!(matches!(ipaddr, IpAddr));
+//! }
+//! ```
+//!
+
+use std::str::Utf8Error;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// An error occured building a `&str` from a C string when
+    /// parsing the name of a interface address instance
+    #[error("Failed to read interface address name. `{0}`")]
+    IntAddrNameParseError(Utf8Error),
+    /// An error ocurred calling `getifaddrs`
+    #[error("Execution of getifaddrs had error result. getifaddrs returned `{0}`")]
+    GetIfAddrsError(i32),
+    /// The current platform is not supported
+    #[error("The current platform `{0}` is not supported")]
+    PlatformNotSupported(String),
+}
+
+#[cfg(target_family = "unix")]
+pub mod unix;
+#[cfg(target_family = "windows")]
+pub mod windows;

```

Comment:
I'm concerned on how this could affect our API usage?

I think that by exporting the module like this we would have something like:

```rust
use local_ip_address::windows::local_ip
```

Instead of:

```rust
use local_ip_address::local_ip
```

As we are taking advantage of conditional compilation, I think we could rely on the following approach:

```suggestion
#[cfg(target_family = "unix")]
pub mod unix::*;

#[cfg(target_family = "windows")]
pub mod windows::*;

```

So, if you are compiling this on `unix`-like systems, you would have `local_ip` implementation for Unix based.
Otherwise you would have for `windows`

Reply:
agreed

Reply:
are you sure that's valid syntax?

Reply:
Nope I just did the suggestion, but I tested this locally:

1. Use `cfg_if` crate: https://crates.io/crates/cfg-if
2. Use `cfg_if::cfg_if!` macro to discriminate on `target_family` CC.

```rust
cfg_if! {    
    if #[cfg(target_family = "unix")] {
        mod unix;
        pub use unix::*;
    }
}

cfg_if! {
    if #[cfg(target_family = "windows")] {
        mod windows;
        pub use windows::*;
    }
}
```

I did a branch pulling from this one to test it locally, this is the only commit ahead: https://github.com/EstebanBorai/local-ip-address/commit/19c9e9fcae09cdea7200e045a1a93a4f1d74e3ca feel free to cherry-pick it or copy from there, sadly I missing the windows test

Reply:
turns out that the only thing missing was de "pub" in front of mod 
`#[cfg(target_family = "unix")]
pub mod unix;
#[cfg(target_family = "unix")]
pub use crate::unix::*;
#[cfg(target_family = "windows")]
pub mod windows;
#[cfg(target_family = "windows")]
pub use crate::windows::*;`

Reply:
I mean no need to add another crate dependency just for this

Reply:
Wohoo!